src/persist/src/operators/mod.rs

```diff
@@ -24,3 +39,121 @@ pub(crate) fn split_ok_err<K, V>(
+///
+/// ```ignore
+/// async fn example(s: Scheduler) {
+///     while scheduler.yield_now().await {
+///     }
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Scheduler(Rc<Cell<bool>>);
+
+/// The future returned by [Scheduler::yield_now]
+#[derive(Debug)]
+pub struct Yield<'a>(&'a Rc<Cell<bool>>);
+
+impl Scheduler {
+    /// Returns a future that will stay pending until timely reschedules the operator
+    pub fn yield_now(&self) -> Yield {
+        Yield(&self.0)
+    }
+}
+
+impl Future for Yield<'_> {
+    type Output = bool;
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
+        if self.0.replace(false) {
+            Poll::Ready(true)
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+trait OpeatorBuilderExt<G: Scope> {
+    /// Creates an operator implementation from supplied async logic constructor.
+    ///
+    /// The logic constructror is expected to return a future that will never return and that
+    /// follows the following pattern:
+    ///
+    /// ```ignore
+    /// op.build_async(scope, move |capabilities, frontier, scheduler| async move {
+    ///     while scheduler.yield_now().await {
+    ///         // async operator logic here
+    ///     }
+    /// });
+    /// ```
+    ///
+    /// Since timely's input handles and frontier notifications are not integrated with the async
+    /// ecosystem the only way to yield control back to timely is by awaiting on
+    /// `scheduler.yield_now()`. The operator will ensure that this call resolves when there is
+    /// more work to do.
+    fn build_async<B, Fut>(self, scope: G, constructor: B)
+    where
+        B: FnOnce(
+            Vec<Capability<G::Timestamp>>,
+            Rc<RefCell<Vec<Antichain<G::Timestamp>>>>,
+            Scheduler,
+        ) -> Fut,
+        Fut: Future + 'static;
+}
+
+impl<G: Scope> OpeatorBuilderExt<G> for OperatorBuilder<G> {
+    fn build_async<B, Fut>(self, scope: G, constructor: B)
+    where
+        B: FnOnce(
+            Vec<Capability<G::Timestamp>>,
+            Rc<RefCell<Vec<Antichain<G::Timestamp>>>>,
+            Scheduler,
+        ) -> Fut,
+        Fut: Future + 'static,
+    {
+        let activator = scope.sync_activator_for(&self.operator_info().address[..]);
+        let waker = futures_util::task::waker(Arc::new(activator));
+        let shared_frontiers = Rc::new(RefCell::new(vec![
+            Antichain::from_elem(Timestamp::minimum());
+            self.shape().inputs()
+        ]));
+
+        self.build(move |capabilities| {
+            let scheduled = Rc::new(Cell::new(false));
+            let mut logic_fut = Box::pin(constructor(
+                capabilities,
+                Rc::clone(&shared_frontiers),
+                Scheduler(Rc::clone(&scheduled)),
+            ));
+            move |frontiers| {
+                // If the future is currently borrowing the shared frontier we'll just go ahead and
```

Comment:
nit: it seems like you poll the future regardless of whether its currently borrowing the shared frontier or not in line 142, but the way the comment is worded it sort of seems like we only poll the future iff its currently borrowing the shared frontier?

Reply:
I'll rephrase the comment but the code is correct. This RefCell borrow can fail if the future also borrowed the frontier internally and kept the borrow alive across an await point. So it is crucial that we poll the future anyway so that it progresses through its state machine and eventually, in some future invocation of the operator, the lock will be released (because the future would have reached the `scheduler.yield_now().await` point

Reply:
@ruchirK I rephrased the comment, let me know if's clear now

Reply:
Yeah the new comment is great! Super minor nit: I think you mean "gets rescheduled" instead of "gets rescheduler", but don't bother fixing that unless you have to go through ci again

Reply:
Trust me I had no doubts the code was correct :)

Reply:
Fixed the typo :) I don't have sympathy for the CI working hard, it can rebuild the whole thing to change a letter :P